lexer.go

```diff
@@ -195,36 +209,65 @@ func (l *Lexer) readrune() (rune, int, error) {
 	}
 
 	_, _ = l.b.WriteRune(rn)
-	return rn, n, nil
+	return rn
+}
+
+// Peek returns the next rune from the buffer without advancing the
+// lexer or underlying reader. The rune stop being valid at the next read
+// call.
+func (l *Lexer) Peek() rune {
+	p := l.PeekN(1)
+	if len(p) < 1 {
+		return EOF
+	}
+	return p[0]
 }
 
-// Peek returns the next n runes from the buffer without advancing the
+// PeekN returns the next n runes from the buffer without advancing the
 // lexer or underlying reader. The runes stop being valid at the next read
-// call. If Peek returns fewer than n runes, it also returns an error
-// indicating why the read is short.
-func (l *Lexer) Peek(n int) ([]rune, error) {
+// call. PeekN may return fewer runes than requested.
+func (l *Lexer) PeekN(n int) []rune {
+	if l.err != nil {
+		return []rune{EOF}
+	}
+
 	p, err := l.r.Peek(n)
-	//nolint:wrapcheck // Error may return io.EOF.
-	return p, err
+	l.setErr(err)
+	if len(p) == 0 {
+		return []rune{EOF}
+	}
+	return p
```

Comment:
Should EOF be appended to the input if at the end of the input? I think not since Peek is trying to support no-copy semantics. But that's inconsistent with returning EOF above.